contracts/oracle/BaseOraclePTPendle.sol

```diff
@@ -0,0 +1,92 @@
+import "pendle/interfaces/IPMarket.sol";
+import { PendlePtOracleLib } from "pendle/oracles/PendlePtOracleLib.sol";
+import "../interfaces/ITreasury.sol";
+
+/// @title BaseOraclePTPendle
+/// @author Angle Labs, Inc.
+/// @notice Base oracle implementation for PT tokens on Pendle
+abstract contract BaseOraclePTPendle {
+    uint256 public constant BASE_18 = 1 ether;
+    uint256 public constant YEAR = 365 days;
+    // @notice The maximum implied rate for the underlying asset, if set well it allows to have a lower bound on the PT token price
+    uint256 public maxImpliedRate;
+    // @notice The duration of the TWAP used to calculate the PT price
+    uint32 public twapDuration;
+
+    /*//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+                                                        EVENTS                                                      
+    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////*/
+
+    event MaxImpliedRateUpdated(uint256 _maxImpliedRate);
+    event TwapPTDurationUpdated(uint256 _twapDuration);
+
+    constructor(uint256 _maxImpliedRate, uint32 _twapDuration) {
+        maxImpliedRate = _maxImpliedRate;
+        twapDuration = _twapDuration;
+    }
+
+    modifier onlyGovernorOrGuardian() {
+        _onlyGovernorOrGuardian();
+        _;
+    }
+
+    function _getQuoteAmount() internal view virtual returns (uint256) {
+        uint256 economicalLowerBound = _economicalPTLowerBoundPrice();
+        uint256 pendlePrice = _pendlePTPrice();
+        uint256 minPrice = economicalLowerBound > pendlePrice ? pendlePrice : economicalLowerBound;
+        uint256 quote = (_detectHackRatio() * minPrice) / BASE_18;
+        return quote;
+    }
+
+    function setMaxImpliedRate(uint256 _maxImpliedRate) external onlyGovernorOrGuardian {
+        maxImpliedRate = _maxImpliedRate;
+        emit MaxImpliedRateUpdated(_maxImpliedRate);
+    }
+
+    function setTwapDuration(uint32 _twapDuration) external onlyGovernorOrGuardian {
+        twapDuration = _twapDuration;
+        emit TwapPTDurationUpdated(_twapDuration);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+                                                       INTERNAL                                                     
+    //////////////////////////////////////////////////////////////////////////////////////////////////////////////////*/
+
+    function _economicalPTLowerBoundPrice() internal view returns (uint256) {
+        uint256 exp = block.timestamp > maturity() ? 0 : maturity() - block.timestamp;
+        if (exp == 0) return BASE_18;
+
+        UD60x18 denominator = UNIT.add(ud(maxImpliedRate)).pow(ud(exp).div(ud(YEAR)));
+        uint256 lowerBound = UNIT.div(denominator).unwrap();
+        return lowerBound;
+    }
```

Comment:
I have some questions here,
If I'm not mistaken, this fonction computes the `ptToAssetPrice` of the `maxImpliedRate following this :
<img width="388" alt="Capture d’écran 2024-04-05 à 16 33 14" src="https://github.com/AngleProtocol/borrow-contracts/assets/83286814/28394be2-6009-4b49-aedb-38b7e4d0cb2f">
So why is `UNIT` added to `ud(maxImpliedRate)).pow(ud(exp).div(ud(YEAR))` ? (it doesn't appear in the formula given by pendle)
And why the result is called lowerBound ? I feel like it's an upper bound given that it is the price for the maximum implied rate

Reply:
This is because the implied rate is` 0.3e17` when it's a `30%` APY. You need to add the one to make you first line correct no?

Reply:
Here I am computing a lower bound because 1/(1+maxRate)**(timeToMaturity/year) is decreasing in `maxRate`.

Economically speaking it is `valeurs actualisées`, normalising the future cash flow (1) from expected rate. Like a zero coupon bond with an interest rate of r%. The cost at t0 of this ZCB is 1/(1+r)**((timeToMaturity-t0)/year).

For the screenshot, this is because `implied rate` is computed like in Uniswap (with cumulatives value).